Improve the message overlay in the 3D editor

Calinou opened the pull request:
This makes the overall experience closer to Blender.

- ~~Use a fixed-width font[^1], fixed decimal count and tab characters to ensure displaying rapidly changing numbers doesn't cause reflowing.~~
  - No longer needed now that the editor defaults to a font with fixed-width digits enabled.
- Add padding for negative numbers to further reduce reflows when a number goes from a negative to a positive value and vice versa.
  - Figure spaces are used to have a space character that has the same width as a minus sign (for fonts with fixed-width digits).
- Improve coordinate display to be more readable.
- Use the `m` suffix for "meter" like in the inspector, and `°` for degrees.
- Use a font outline to make the text more readable on bright backgrounds.
- Use sentence case and don't end messages with a period.

These changes can be ported to the 2D editor once we reach an agreement on them.

- This closes https://github.com/godotengine/godot-proposals/issues/15012 (by providing a similar solution that improves readability on mixed-color backgrounds).

[^1]: If we used a default editor font with fixed-width numbers, we could keep using a proportional font for letters. Many other places in the editor would benefit from this, including the View Information/View Frame Time panels and the inspector. This exists as an OpenType stylistic alternate in some fonts like [Inter](https://rsms.me/inter/), but Noto Sans does not have this alternate.

## Preview

<img width="501" height="89" alt="image" src="https://github.com/user-attachments/assets/45052b31-93cc-4f20-b355-f34410f85825" />

<img width="429" height="92" alt="image" src="https://github.com/user-attachments/assets/a9702229-f1fe-4013-8fba-0fd3540f8317" />

<details>
<summary>Old version</summary>

https://github.com/user-attachments/assets/096332cc-aed6-4015-be91-adaca5081490

![image](https://github.com/user-attachments/assets/8088f4c3-f41a-4a69-812f-95ef52be180d)

![image](https://github.com/user-attachments/assets/b2b9c35a-00a9-49e0-8928-031383996bc2)

![image](https://github.com/user-attachments/assets/dafd8fd1-9c72-4a3d-86fe-269ca1392c8f)
</summary>

Calinou committed:
Improve the message overlay in the 3D editor

This makes the overall experience closer to Blender.

- Use a fixed-width font, fixed decimal count and tab characters
  to ensure displaying rapidly changing numbers doesn't cause reflowing.
- Add padding for negative numbers to further reduce reflows when a
  number goes from a negative to a positive value and vice versa.
- Improve coordinate display to be more readable.
- Use the `m` suffix for "meter" like in the inspector, and `°` for degrees.
- Use a font outline to make the text more readable on bright backgrounds.
- Use sentence case and don't end messages with a period.

passivestar commented:
I like sentence case and suffixes but I'm not sure about other changes. There's too much padding and too much contrast. Comparison with blender:

![111](https://github.com/user-attachments/assets/e8e97e16-ceeb-48f3-9ec3-b4f12dd6f8f6)

Paddings make numbers clump with wrong labels ("0.000m" is visually closer to "Z" than to "Y"). It would make sense to leave space for at most one extra character (`-`), and you'll need extra space after each component to better separate them, like what blender does. There's also no space before the total distance rn

As for contrast `#fff` and `#000` is just too much. Readability is important but on bright background pure black `#000` outline has the opposite effect and makes it somewhat harder to read, imo it shouldn't be pure black. Semitransparent maybe

Calinou commented:
@passivestar I've pushed an updated version with improved padding and reduced contrast for the outline. See OP for updated screenshots.

Mickeon commented on editor/plugins/gizmos/visible_on_screen_notifier_3d_gizmo_plugin.cpp:
```diff
@@ -64,11 +64,11 @@ String VisibleOnScreenNotifier3DGizmoPlugin::get_handle_name(const EditorNode3DG
-			return "Pos X";
+			return "Position X";
 		case 4:
-			return "Pos Y";
+			return "Position Y";
 		case 5:
-			return "Pos Z";
+			return "Position Z";
```
All of these are not going through `TTR()`, is this normal?

Calinou replied:
Good catch, these should be using `TTR()`.

Mickeon commented on editor/plugins/gizmos/audio_stream_player_3d_gizmo_plugin.cpp:
```diff
@@ -61,7 +61,7 @@ int AudioStreamPlayer3DGizmoPlugin::get_priority() const {
 
 String AudioStreamPlayer3DGizmoPlugin::get_handle_name(const EditorNode3DGizmo *p_gizmo, int p_id, bool p_secondary) const {
-	return "Emission Radius";
+	return "Emission radius";
```
This is also left not localized for some reason, in retrospective.

Mickeon commented on doc/classes/EditorNode3DGizmo.xml:
```diff
@@ -43,7 +43,7 @@
 			<param index="1" name="secondary" type="bool" />
 			<description>
-				Override this method to return the name of an edited handle (handles must have been previously added by [method add_handles]). Handles can be named for reference to the user when editing.
+				Override this method to return the name of an edited handle (handles must have been previously added by [method add_handles]). Handles can be named for reference to the user when editing. By convention, handle names should use [i]Sentence case[/i] (e.g. [code]"Inner radius"[/code] instead of [code]"Inner Radius"[/code]).
```
Genuinely the first time "Sentence case" would be ever referenced in the documentation. I would consider spelling out what it means, but if it sounds wrong, I guess this is okay.
This remark also fits quite snuggly as a note.
```suggestion
				Override this method to return the name of an edited handle (handles must have been previously added by [method add_handles]). Handles can be named for reference to the user when editing.
				[b]Note:[/b] By convention, the first word of the name should be capitalized, while the rest should be in lowercase. For example, [code]"Inner radius"[/code].
```
Or similar?

Mickeon reviewed (commented):
Random addendum: This PR would further hammer in that 3D uses meters. There's some bits of the class reference I've seen that still assume the unit of measurement is generic (as was the case in `3.x`).

bruvzg commented:
> Should we have a fallback to handle thin (possibly non-breaking) spaces?

NBSP should be already supported (in terms of line breaking). Fallback server simply pulls glyph advances from the font. I guess we can add an hardcoded `x * normal_space` overrides for the special spaces.

bruvzg commented:
> I guess we can add an hardcoded x * normal_space overrides for the special spaces.

```diff
diff --git a/modules/text_server_fb/text_server_fb.cpp b/modules/text_server_fb/text_server_fb.cpp
index 5828991321..f72b154597 100644
--- a/modules/text_server_fb/text_server_fb.cpp
+++ b/modules/text_server_fb/text_server_fb.cpp
@@ -4754,7 +4754,8 @@ bool TextServerFallback::_shaped_text_shape(const RID &p_shaped) {
 				gl.end = j + 1;
 				gl.count = 1;
 				gl.font_size = span.font_size;
-				gl.index = (int32_t)sd->text[j - sd->start]; // Use codepoint.
+				int32_t codepoint = (int32_t)sd->text[j - sd->start];
+				gl.index = codepoint; // Use codepoint.
 				if (gl.index == 0x0009 || gl.index == 0x000b) {
 					gl.index = 0x0020;
 				}
@@ -4786,9 +4787,25 @@ bool TextServerFallback::_shaped_text_shape(const RID &p_shaped) {
 				if (gl.font_rid.is_valid()) {
 					double scale = _font_get_scale(gl.font_rid, gl.font_size);
 					bool subpos = (scale != 1.0) || (_font_get_subpixel_positioning(gl.font_rid) == SUBPIXEL_POSITIONING_ONE_HALF) || (_font_get_subpixel_positioning(gl.font_rid) == SUBPIXEL_POSITIONING_ONE_QUARTER) || (_font_get_subpixel_positioning(gl.font_rid) == SUBPIXEL_POSITIONING_AUTO && gl.font_size <= SUBPIXEL_POSITIONING_ONE_HALF_MAX_SIZE);
-					if (sd->text[j - sd->start] != 0 && !is_linebreak(sd->text[j - sd->start])) {
+					if (codepoint != 0 && !is_linebreak(codepoint)) {
 						if (sd->orientation == ORIENTATION_HORIZONTAL) {
-							gl.advance = _font_get_glyph_advance(gl.font_rid, gl.font_size, gl.index).x;
+							if (codepoint == 0x2000 || codepoint == 0x2002) {
+								gl.advance = gl.font_size * 0.5;
+							} else if (codepoint == 0x2001 || codepoint == 0x2003) {
+								gl.advance = gl.font_size;
+							} else if (codepoint == 0x2004) {
+								gl.advance = gl.font_size * 0.33333333333;
+							} else if (codepoint == 0x2005) {
+								gl.advance = gl.font_size * 0.25;
+							} else if (codepoint == 0x2006) {
+								gl.advance = gl.font_size * 0.16666666666;
+							} else if (codepoint == 0x2009 || codepoint == 0x202f) {
+								gl.advance = gl.font_size * 0.2;
+							} else if (codepoint == 0x200a) {
+								gl.advance = gl.font_size * 0.125;
+							} else {
+								gl.advance = _font_get_glyph_advance(gl.font_rid, gl.font_size, gl.index).x;
+							}
 							gl.x_off = 0;
 							gl.y_off = _font_get_baseline_offset(gl.font_rid) * (double)(_font_get_ascent(gl.font_rid, gl.font_size) + _font_get_descent(gl.font_rid, gl.font_size));
 							sd->ascent = MAX(sd->ascent, _font_get_ascent(gl.font_rid, gl.font_size) + _font_get_spacing(gl.font_rid, SPACING_TOP));
@@ -4803,7 +4820,7 @@ bool TextServerFallback::_shaped_text_shape(const RID &p_shaped) {
 					}
 					if (j < sd->end - 1) {
 						// Do not add extra spacing to the last glyph of the string.
-						if (is_whitespace(sd->text[j - sd->start])) {
+						if (is_whitespace(codepoint)) {
 							gl.advance += sd->extra_spacing[SPACING_SPACE] + _font_get_spacing(gl.font_rid, SPACING_SPACE);
 						} else {
 							gl.advance += sd->extra_spacing[SPACING_GLYPH] + _font_get_spacing(gl.font_rid, SPACING_GLYPH);
```

Something like this seems to give results (right) similar to `TextServerAdvanced` (left), but not sure if extra handling should be added to fallback text server, it intended to me minimal.
<img width="437" alt="Screenshot 2025-05-15 at 10 40 48" src="https://github.com/user-attachments/assets/8a769bcc-b6df-4f00-a229-52355cee5a5a" />

Calinou commented:
Rebased and tested again, it works as expected.

I switched from the fixed-width font back to the default editor font, but with *figure spaces* as stand-ins for the `-` sign. This space character has the same width as a digit (or minus sign) in fonts with fixed-width digits, which includes the default editor font.

passivestar commented:
It's in a better shape but I'd still suggest tighter axes grouping, both because it's easier to parse and because when there's a lot of space the string as a whole just feels too big aesthetically compared to the rest of the editor UI text. I think no space between the number and `m` is fine

PR:

<img width="834" height="133" alt="a" src="https://github.com/user-attachments/assets/514b54dc-adc4-4cee-97ef-5f8b6013d4fd" /><br>

My suggestion:

<img width="834" height="133" alt="b" src="https://github.com/user-attachments/assets/94af42d4-6dfe-4717-9932-2806d8c96ad3" /><br>
